haskell/cabal.bzl

```diff
@@ -671,53 +698,118 @@ stack_snapshot = repository_rule(
-
-  Alternatively
-  ```bzl
-  stack_snapshot(
-      name = "stackage",
-      packages = ["conduit", "lens", "zlib"],
-      tools = ["@happy//:happy", "@c2hs//:c2hs"],
-      local_Snapshot = "//:snapshot.yaml",
-      deps = ["@zlib.dev//:zlib"],
-  ```
-  Does the same as the previous example, provided there is a
-  `snapshot.yaml`, at the root of the repository with content
-  ```yaml
-  resolver: lts-13.15
-
-  packages:
-    - zlib-0.6.2
-  ```
-
-This rule will use Stack to compute the transitive closure of the
-subset of the given snapshot listed in the `packages` attribute, and
-generate a dependency graph. If a package in the closure depends on
-system libraries or other external libraries, use the `deps` attribute
-to list them. This attribute works like the
-`--extra-{include,lib}-dirs` flags for Stack and cabal-install do.
-
-Packages that are in the snapshot need not have their versions
-specified. But any additional packages or version overrides will have
-to be specified with a package identifier of the form
-`<package>-<version>` in the `packages` attribute.
-
-In the external repository defined by the rule, all given packages are
-available as top-level targets named after each package.
+def _get_platform(repository_ctx):
+    """Map OS name and architecture to Stack platform identifiers."""
+    os_name = repository_ctx.os.name.lower()
+    if os_name.startswith("linux"):
+        os = "linux"
+    elif os_name.startswith("mac os"):
+        os = "osx"
+    elif os_name.find("freebsd") != -1:
+        os = "freebsd"
+    elif os_name.find("windows") != -1:
+        os = "windows"
+
+    result = repository_ctx.execute(["uname", "-m"])
+    if result.stdout.strip() in ["arm", "armv7l"]:
+        arch = "arm"
+    elif result.stdout.strip() in ["aarch64"]:
+        arch = "aarch64"
+    elif result.stdout.strip() in ["amd64", "x86_64", "x64"]:
+        arch = "x86_64"
+    elif result.stdout.strip() in ["i386", "i486", "i586", "i686"]:
+        arch = "i386"
+
+    return (os, arch)
+
+def _fetch_stack_impl(repository_ctx):
+    repository_ctx.file("BUILD.bazel")
+    stack_cmd = repository_ctx.which("stack")
+    if stack_cmd:
+        exec_result = _execute_or_fail_loudly(repository_ctx, [stack_cmd, "--version"])
+        stack_version_words = exec_result.stdout.split(" ")
+
+        # Stack version strings are sometimes prefixed by the word
+        # "Version", sometimes not.
+        if stack_version_words[0] == "Version":
+            stack_major_version = int(stack_version_words[1].split(".")[0])
+        else:
+            stack_major_version = int(stack_version_words[0].split(".")[0])
+        if stack_major_version < 2:
+            print("Stack version not recent enough. Downloading a newer version...")
+        else:
+            repository_ctx.symlink(stack_cmd, "stack")
+            return
+
+    # If we can't find Stack, download it.
+    (os, arch) = _get_platform(repository_ctx)
```

Comment:
Should we detect NixOS and fail gracefully in that case? The bindist will probably fail with the usual and quite misleading "no such file or directory" error.

Reply:
Yeah, we should put this behind a flag and only enable it for bindists

Reply:
I don't know how to reliably detect NixOS. Maybe we could improve the error message in this case with hints? Being on NixOS will be the common case for why the binary failed...